devicetypes/stelpro/stelpro-ki-zigbee-thermostat.src/stelpro-ki-zigbee-thermostat.groovy

```diff
@@ -0,0 +1,667 @@
+	if (settings.zipcode) {
+		log.debug "ZipCode: ${settings.zipcode}"
+		weather = getWeatherFeature( "conditions", settings.zipcode )
+
+		// Check if the variable is populated, otherwise return.
+		if (!weather) {
+			log.debug( "Something went wrong, no data found." )
+			return false
+		}
+		
+		def tempToSend
+		
+		if(getTemperatureScale() == "C" ) {
+			tempToSend = weather.current_observation.temp_c
+			log.debug( "Outdoor Temperature: ${tempToSend} C" )
+		}
+		else {
+			tempToSend = weather.current_observation.temp_f
+			log.debug( "Outdoor Temperature: ${tempToSend} F" )
+		}	   
+		sendEvent( name: 'outsideTemp', value: tempToSend )
+		quickSetOutTemp(tempToSend)
+	}
+}
+
+def scheduledUpdateWeather() {
+	def actions = updateWeather()
+
+	if (actions) {
+		sendHubCommand(actions)
+	}
+}
+
+def getModeMap() { [
+	"00":"off",
+	"04":"heat",
+	"05":"eco"
+]}
+
+def getFanModeMap() { [
+	"04":"fanOn",
+	"05":"fanAuto"
+]}
+
+/**
+  * PING is used by Device-Watch in attempt to reach the Device
+**/
+def ping() {
+	zigbee.readAttribute(0x201, 0x0000)
+}
+
+def poll() {
+	sendEvent( name: 'change', value: 0 )
+	delayBetween([
+			updateWeather(),
+			zigbee.readAttribute(0x201, 0x0000),	//Read Local Temperature
+			zigbee.readAttribute(0x201, 0x0008),	//Read PI Heating State
+			zigbee.readAttribute(0x201, 0x0012),	//Read Heat Setpoint
+			zigbee.readAttribute(0x201, 0x001C),	//Read System Mode
+			zigbee.readAttribute(0x201, 0x401C, ["mfgCode": "0x1185"]),	//Read Manufacturer Specific Setpoint Mode
+			zigbee.readAttribute(0x204, 0x0000),	//Read Temperature Display Mode
+			zigbee.readAttribute(0x204, 0x0001)		//Read Keypad Lockout
+		], 200)
+}
+
+
+
+def getTemperature(value) {
+	if (value != null) {
+		log.debug("value $value")
+		def celsius = Integer.parseInt(value, 16) / 100
+		if (getTemperatureScale() == "C") {
+			return celsius
+		}
+		else {
+			return Math.round(celsiusToFahrenheit(celsius))
+		}
+	}
+}
+
+def refresh() {
+	poll()
+}
+
+def quickSetHeat(degrees) {
+	sendEvent( name: 'change', value: 1 )
+	setHeatingSetpoint(degrees, 0)
+}
+
+def setHeatingSetpoint(preciseDegrees, delay = 0) {
+	if (preciseDegrees != null) {
+		def temperatureScale = getTemperatureScale()
+		def degrees = new BigDecimal(preciseDegrees).setScale(1, BigDecimal.ROUND_HALF_UP)
+
+		log.debug "setHeatingSetpoint({$degrees} ${temperatureScale})"
+		
+		sendEvent(name: "heatingSetpoint", value: degrees, unit: temperatureScale, data: [heatingSetpointRange: heatingSetpointRange])
+		sendEvent(name: "thermostatSetpoint", value: degrees, unit: temperatureScale, data: [thermostatSetpointRange: thermostatSetpointRange])
+		
+		def celsius = (getTemperatureScale() == "C") ? degrees : (fahrenheitToCelsius(degrees) as Float).round(2)
+		delayBetween([
+			zigbee.writeAttribute(0x201, 0x12, 0x29, hex(celsius * 100)),
+			zigbee.readAttribute(0x201, 0x12),	//Read Heat Setpoint
```

Comment:
Typically we write the value at the width of the field it is filling.  I.e. `0x0201` and `0x0012`, but it's not a big deal.